src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs

```diff
@@ -100,21 +100,71 @@ internal static IReadOnlyList<PackageIdentity> UpdateTransitiveDependencies(Lock
 
         private static NuGetVersion GetInstalledVersion(string libraryName, NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets)
         {
-            return targets
-                ?.FirstOrDefault(t => t.TargetFramework.Equals(targetFramework) && string.IsNullOrEmpty(t.RuntimeIdentifier))
-                ?.Libraries
-                ?.FirstOrDefault(a => a.Name.Equals(libraryName, StringComparison.OrdinalIgnoreCase))
-                ?.Version;
+            // PERF: Intentionally avoiding LINQ and foreach to avoid allocating capture classes and enumerators
+            if (targets != null)
+            {
+                LockFileTarget target = default;
+                for (int i = 0; i < targets.Count; ++i)
+                {
+                    LockFileTarget t = targets[i];
+                    if (t.TargetFramework.Equals(targetFramework) && string.IsNullOrEmpty(t.RuntimeIdentifier))
+                    {
+                        target = t;
+                    }
+                }
+
+                if (target != null && target.Libraries != null)
+                {
+                    for (int i = 0; i < target.Libraries.Count; ++i)
+                    {
+                        LockFileTargetLibrary a = target.Libraries[i];
+                        if (a != null && a.Name.Equals(libraryName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return a.Version;
+                        }
+                    }
+                }
+            }
+
+            return null;
         }
 
         private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(LockFileTargetLibrary library, NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets)
         {
-            return targets
-                ?.FirstOrDefault(t => t.TargetFramework.Equals(targetFramework) && string.IsNullOrEmpty(t.RuntimeIdentifier))
-                ?.Libraries
-                ?.Where(lib => lib.Name.Equals(library.Name, StringComparison.OrdinalIgnoreCase))
-                ?.SelectMany(lib => lib.Dependencies)
-                ?.ToList();
+            // PERF: Intentionally avoiding LINQ and foreach to avoid allocating capture classes and enumerators
+            if (targets != null)
+            {
+                LockFileTarget target = default;
+                for (int i = 0; i < targets.Count; ++i)
+                {
+                    LockFileTarget t = targets[i];
+                    if (t.TargetFramework.Equals(targetFramework) && string.IsNullOrEmpty(t.RuntimeIdentifier))
+                    {
+                        target = t;
+                        break;
```

Comment:
There's a break here, but not one above on line 112. 

This is guaranteed to always have 1 hit max though, so the break probably doesn't do much.

Reply:
Nice catch. I'll update it just in case that guarantee happens to change in the future.